src/Core/OrganizationFeatures/AuthorizationHandlers/GroupAuthorizationHandler.cs

```diff
@@ -0,0 +1,113 @@
+
+            case not null when requirement == GroupOperations.Read:
+                await CanReadAsync(context, requirement, resource);
+                break;
+
+            case not null when requirement == GroupOperations.Update:
+                await CanUpdateAsync(context, requirement, resource);
+                break;
+
+            case not null when requirement == GroupOperations.Delete:
+                await CanDeleteAsync(context, requirement, resource);
+                break;
+        }
+    }
+
+    protected override async Task HandleRequirementAsync(AuthorizationHandlerContext context,
+        GroupOperationRequirement requirement)
+    {
+        switch (requirement)
+        {
+            case not null when requirement.Name == nameof(GroupOperations.ReadAllForOrganization):
+                await ReadAllGroupsAsync(context, requirement);
+                break;
+        }
+    }
+
+    private async Task CanCreateAsync(AuthorizationHandlerContext context, GroupOperationRequirement requirement,
+        Group resource) => await CanManageAsync(context, requirement, resource);
+
+    private async Task CanReadAsync(AuthorizationHandlerContext context, GroupOperationRequirement requirement,
+        Group resource) => await CanManageAsync(context, requirement, resource);
+
+    private async Task CanUpdateAsync(AuthorizationHandlerContext context, GroupOperationRequirement requirement,
+        Group resource) => await CanManageAsync(context, requirement, resource);
+
+    private async Task CanDeleteAsync(AuthorizationHandlerContext context, GroupOperationRequirement requirement,
+        Group resource) => await CanManageAsync(context, requirement, resource);
+
+    private async Task CanManageAsync(AuthorizationHandlerContext context, GroupOperationRequirement requirement,
+        Group resource)
+    {
+        var org = _currentContext.GetOrganization(resource.OrganizationId);
+        if (org == null)
+        {
+            context.Fail();
+            return;
+        }
+
+        var success = org.Type == OrganizationUserType.Owner ||
+                      org.Type == OrganizationUserType.Admin ||
+                      org.Permissions.ManageGroups ||
+                      await _currentContext.ProviderUserForOrgAsync(org.Id);
+
+        if (success)
+        {
+            context.Succeed(requirement);
+        }
+    }
+
+    private async Task ReadAllGroupsAsync(AuthorizationHandlerContext context,
+        GroupOperationRequirement requirement)
+    {
+        var org = _currentContext.GetOrganization(requirement.OrganizationId.GetValueOrDefault());
+        if (org == null)
+        {
+            context.Fail();
+            return;
+        }
+
+        var success = org.Type == OrganizationUserType.Owner ||
+                      org.Type == OrganizationUserType.Admin ||
+                      org.Type == OrganizationUserType.Manager ||
+                      org.Permissions.ManageGroups ||
```

Comment:
`Permissions` only exists when `org.Type == OrganizationUserType.Custom` otherwise its null and will throw an exception here. (I just ran into this myself)

Probably easiest to just `&&` all of the `org.Permissions.*` checks with a null check when there are this many permissions to check.

Reply:
I instantiated a default `Permissions` object on `CurrentContentOrganization` so it'll never be null :) no null checks required!